include/circt/Dialect/SMT/SMTOps.td

```diff
@@ -16,8 +16,152 @@ include "mlir/IR/EnumAttr.td"
 include "mlir/IR/OpAsmInterface.td"
 include "mlir/Interfaces/InferTypeOpInterface.td"
 include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/Interfaces/ControlFlowInterfaces.td"
 
 class SMTOp<string mnemonic, list<Trait> traits = []> :
   Op<SMTDialect, mnemonic, traits>;
 
+def DeclareConstOp : SMTOp<"declare_const", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>
+]> {
+  let summary = "declare a symbolic value of a given sort";
+  let description = [{
+    This operation declares a symbolic value just as the `declare-const`
+    statement in SMT-LIB 2.6. The result type determines the SMT sort of the
+    symbolic value. The returned value can then be used to refer to the symbolic
+    value instead of using the identifier like in SMT-LIB.
+
+    The optionally provided string will be used as a prefix for the newly
+    generated identifier (useful for easier readability when exporting to
+    SMT-LIB). Each `declare_const` will always provide a unique new symbolic
+    value even if the identifier strings are the same.
+
+    Note that this operation cannot be marked as Pure since two operations (even
+    with the same identifier string) could then be CSEd, leading to incorrect
+    behavior.
+  }];
+
+  let arguments = (ins OptionalAttr<StrAttr>:$namePrefix);
+  let results = (outs Res<AnySMTType, "a symbolic value", [MemAlloc]>:$result);
+
+  let assemblyFormat = [{
+    ($namePrefix^)? attr-dict `:` qualified(type($result))
+  }];
+}
+
+def SolverOp : SMTOp<"solver", [
+  IsolatedFromAbove,
+  SingleBlockImplicitTerminator<"smt::YieldOp">,
+]> {
+  let summary = "create a solver instance within a lifespan";
+  let description = [{
+    This operation defines an SMT context with a solver instance. SMT operations
+    are only valid when being executed between the start and end of the region
+    of this operation. Any invocation outside is undefined. However, they do not
+    have to be direct children of this operation. For example, it is allowed to
+    have SMT operations in a `func.func` which is only called from within this
+    region. No SMT value may enter or exit the lifespan of this region (such
+    that no value created from another SMT context can be used in this scope and
+    the solver can deallocate all state required to keep track of SMT values at
+    the end).
+
+    As a result, the region is comparable to an entire SMT-LIB script, but
+    allows for concrete operations and control-flow. Concrete values may be
+    passed in and returned to influence the computations after the `smt.solver`
+    operation.
+
+    Example:
+    ```mlir
+    %0:2 = smt.solver (%in) {smt.some_attr} : (i8) -> (i8, i32) {
```

Comment:
What does the input/output correspond to?

Reply:
The body of the `smt.solver` is isolated from above. The reason is that SMT constants cannot be hoisted out of the solver and we have a bit more leverage to check that no SMT value from another solver is used in this solver. Both of which is a big problem in the LLVM lowering. The earlier because no smt context is initialized outside the solver's body and just attempting to build an expression node would lead to a crash, the latter because the nodes would have been built in different contexts with potentially other configs, i.e., are not compatible.
Still, it's useful to be able to pass in concrete values to influence the construction of the SMT expressions. And the output is useful to report the result of the solver (was is sat or unsat, a counter example, etc.). Since they are just passthrough values, they must match the block arguments and yield operands respectively.